core/src/replay_stage.rs

```diff
@@ -3349,16 +3357,72 @@ impl ReplayStage {
                     duplicate_slots_tracker,
                     duplicate_confirmed_slots,
                     epoch_slots_frozen_slots,
+                    slow_slots_by_expiry,
                     ..
                 }) = &mut tbft_structs
                 {
+                    let block = (bank.slot(), bank.hash());
                     // Needs to be updated before `check_slot_agrees_with_cluster()` so that
                     // any updates in `check_slot_agrees_with_cluster()` on fork choice take
                     // effect
-                    heaviest_subtree_fork_choice.add_new_leaf_slot(
-                        (bank.slot(), bank.hash()),
-                        Some((bank.parent_slot(), bank.parent_hash())),
-                    );
+                    heaviest_subtree_fork_choice
+                        .add_new_leaf_slot(block, Some((bank.parent_slot(), bank.parent_hash())));
+
+                    // Check if this block took too long, if so mark it as invalid.
+                    // This prevents voting or building on any blocks in this fork.
+                    // If the fork reaches > 52% votes (DC) or we freeze a bank after SLOW_SLOT_LOCKOUT
+                    // slots we expire this invalid penality.
+                    //
+                    // Notably in slow slot scenarios, a block cannot be marked as invalid
+                    // *AFTER* we have voted for it, avoiding the hairy scenarios involved with
+                    // switching to a lower staked fork in duplicate block scenarios
+                    let (current_tick_height, timeout_grace_ticks) = {
+                        let poh_recorder_r = poh_recorder.read().unwrap();
+                        let ticks_per_slot = poh_recorder_r.ticks_per_slot();
+
+                        (
+                            poh_recorder_r.tick_height(),
+                            (ticks_per_slot as f64 * SLOT_TIMEOUT_TICK_FACTOR) as u64,
```

Comment:
can make this dynamic based on root slot distance as well, if we're having trouble maybe increase the timeout.